tiledb/sm/query/query_condition.cc

```diff
@@ -51,158 +54,77 @@ QueryCondition::QueryCondition() {
 }
 
 QueryCondition::QueryCondition(const QueryCondition& rhs)
-    : clauses_(rhs.clauses_)
-    , combination_ops_(rhs.combination_ops_) {
+    : tree_(rhs.tree_ == nullptr ? nullptr : rhs.tree_->clone()) {
 }
 
 QueryCondition::QueryCondition(QueryCondition&& rhs)
-    : clauses_(std::move(rhs.clauses_))
-    , combination_ops_(std::move(rhs.combination_ops_)) {
+    : tree_(std::move(rhs.tree_)) {
 }
 
 QueryCondition::~QueryCondition() {
 }
 
 QueryCondition& QueryCondition::operator=(const QueryCondition& rhs) {
-  clauses_ = rhs.clauses_;
-  combination_ops_ = rhs.combination_ops_;
+  if (this != &rhs) {
+    tree_ = rhs.tree_ == nullptr ? nullptr : rhs.tree_->clone();
+  }
 
   return *this;
 }
 
 QueryCondition& QueryCondition::operator=(QueryCondition&& rhs) {
-  clauses_ = std::move(rhs.clauses_);
-  combination_ops_ = std::move(rhs.combination_ops_);
-
+  tree_ = std::move(rhs.tree_);
   return *this;
 }
 
 Status QueryCondition::init(
     std::string&& field_name,
     const void* const condition_value,
     const uint64_t condition_value_size,
-    const QueryConditionOp op) {
-  if (!clauses_.empty()) {
+    const QueryConditionOp& op) {
+  if (tree_) {
     return Status_QueryConditionError("Cannot reinitialize query condition");
   }
 
-  clauses_.emplace_back(
-      std::move(field_name), condition_value, condition_value_size, op);
+  // AST Construction.
+  tree_ = tdb_unique_ptr<ASTNode>(tdb_new(
```

Comment:
I've thought about this too. It should be quick but I'd prefer to not attack it in this PR because there are too many moving parts and it also requires an API change. There is also already a factory create function that handles this case. I think the aim will be to probably eliminate the C style interface for the most part in the cpp API but that is something that definitely could break backwards compatibility. @Shelnutt2 @ihnorton  thoughts?

Reply:
Ok, let's make sure there is a story about that.